pyteal/ast/opup.py

```diff
@@ -0,0 +1,163 @@
+            if target_app_id is None:
+                raise TealInputError(
+                    "target_app_id must be specified in Explicit OpUp mode"
+                )
+            require_type(target_app_id, TealType.uint64)
+            self.target_app_id = target_app_id
+        elif mode == OpUpMode.OnCall:
+            self.target_app_id_slot = ScratchSlot()
+        else:
+            raise TealInputError("Invalid OpUp mode provided")
+
+        self.mode = mode
+
+        # A budget buffer is necessary to deal with an edge case of ensure_budget():
+        #   if the current budget is equal to or only slightly higher than the
+        #   required budget then it's possible for ensure_budget() to return with a
+        #   current budget less than the required budget. The buffer prevents this
+        #   from being the case.
+        self.buffer = Int(50)
+
+    def _create_app(self) -> Expr:
+        self.target_app_id = self.target_app_id_slot.load()
+        return Seq(
+            InnerTxnBuilder.Begin(),
+            InnerTxnBuilder.SetFields(
+                {
+                    TxnField.type_enum: TxnType.ApplicationCall,
+                    TxnField.approval_program: ON_CALL_APP,
+                    TxnField.clear_state_program: ON_CALL_APP,
+                }
+            ),
+            InnerTxnBuilder.Submit(),
+            self.target_app_id_slot.store(Gitxn[0].created_application_id()),
+        )
+
+    def _delete_app(self, app_id: Expr) -> Expr:
+        return Seq(
+            InnerTxnBuilder.Begin(),
+            InnerTxnBuilder.SetFields(
+                {
+                    TxnField.type_enum: TxnType.ApplicationCall,
+                    TxnField.on_completion: OnComplete.DeleteApplication,
+                    TxnField.application_id: app_id,
+                }
+            ),
+            InnerTxnBuilder.Submit(),
+        )
+
+    def _ensure_budget_expr(self, required_budget: Expr) -> Expr:
+        return While(Global.opcode_budget() < required_budget).Do(
+            Seq(
+                InnerTxnBuilder.Begin(),
+                InnerTxnBuilder.SetFields(
+                    {
+                        TxnField.type_enum: TxnType.ApplicationCall,
+                        TxnField.application_id: self.target_app_id,
+                    }
+                ),
+                InnerTxnBuilder.Submit(),
+            )
+        )
+
+    def ensure_budget(self, required_budget: Expr) -> Expr:
+        """Ensure that the budget will be at least the required_budget."""
+        require_type(required_budget, TealType.uint64)
+
+        buffered_budget = required_budget + self.buffer
+        if self.mode == OpUpMode.OnCall:
+            return If(Global.opcode_budget() < buffered_budget).Then(
+                Seq(
+                    self._create_app(),
+                    self._ensure_budget_expr(buffered_budget),
+                    self._delete_app(self.target_app_id),
```

Comment:
@algoidurovic Gauging my understanding - Is the following statement accurate?

If `_ensure_budget_expr` fails, then the created app lives indefinitely.

_If_ my understanding is accurate, I'm wondering if it's possible to avoid leaking the app by creating an inner transaction group via `InnerTxnBuilder` containing (1) create app, (2) op up, and (3) delete app?

Reply:
Had a verbal with @algoidurovic - I misunderstood inner transaction semantics.  Any failing inner transaction (regardless of grouping) fails the app invocation and all associated transactions. 

So - The PR does _not_ dangle app IDs.  Sorry for the confusion.  Resolving.